contracts/src/components/stake.cairo

```
@@ -90,65 +93,87 @@ mod StakeComponent {
+                } else {
+                    panic!("Attempting to discount more than what's staked");
+                }
+            };
+        }
 
-                    let balance = payable.balance_of(*token_used, get_contract_address());
+        fn __generate_token_ratio(
+            ref self: ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            let existing_ratio = self.token_ratios.read(token);
+            if existing_ratio != 0 {
+                return existing_ratio;
+            }
 
-                    let total_staked = get_total_stake_for_token(ref stake_totals, token_key);
+            let mut payable = get_dep_component_mut!(ref self, Payable);
+            let total_staked = self.token_stakes.read(token);
+            let balance = payable.balance_of(token, get_contract_address());
+            let ratio = calculate_refund_ratio(total_staked, balance);
+            self.token_ratios.write(token, ratio);
+            ratio
+        }
 
-                    let total_tax = get_total_tax_for_token(ref tax_totals, token_key);
+        fn __process_refund(
+            ref self: ComponentState<TContractState>,
+            mut land: Land,
+            mut store: Store,
+            refund_amount: u256
+        ) {
+            let mut payable = get_dep_component_mut!(ref self, Payable);
 
-                    let ratio = calculate_refund_ratio(total_staked + total_tax, balance);
+            let validation_result = payable
+                .validate(land.token_used, get_contract_address(), refund_amount);
+            let status = payable.transfer(land.owner, validation_result);
+            assert(status, errors::ERC20_REFUND_FAILED);
 
-                    token_ratios.insert(token_key, NullableTrait::new(ratio));
-                };
+            let current_total = self.token_stakes.read(land.token_used);
+            let new_total = if current_total > refund_amount {
```

If I understand correctly, in the event that the computations were wrong, the last people in the list won't get their money back, right?

---

No, we do the ratio based on what's in our balance sheets and what we should return, but with that ratio, everyone should receive money accordingly

I did that validation because I'm used to handling subtractions that way with Cairo for negative number errors

---

That's what I suspected! IMO because the contracts are upgradable, I would have thrown an exception instead of silently failing